### DISCUSSION TITLE
How to forward `ZephyrConfig.cmake` location to Sub-Builds in `multi_image.cmake`

### DISCUSSION THREAD

**Discussion by TorstenRobitzki** [zephyrproject-rtos/zephyr#70762] [Q&A]
Hello,
so far my Zephyr build works for my first project. Now, I want to add a bootloader (MCUboot). I've included `ZephyrConfig.cmake` into CMake, by loading the module with a hint:

    find_package(Zephyr REQUIRED HINTS ${CMAKE_CURRENT_LIST_DIR}/../zephyr)

    project(firmware
        VERSION 0.0.0
        LANGUAGES C CXX ASM)

This way, I get always the version of Zephyr, which was checked out by west in a relative position to the current checkout of the project.

Now, I've enabled MCUboot, by adding `CONFIG_BOOTLOADER_MCUBOOT=y` to my project configuration. In `multi_image.cmake` is a recursive invocation of CMake, which then opens the CMakeLists.txt of MCUboot, which tries to load the Zephyr module by:

    find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
    project(NONE)

If that variable is requires, why isn't the Zephyr build system not setting that variable, when spawning subprocesses? Any idea, how to work around this, without redundantly setting the environment variable?

**Comment by tejlmand** [comment]
>  I've included ZephyrConfig.cmake into CMake, by loading the module with a hint:
> ...
> This way, I get always the version of Zephyr, which was checked out by west in a relative position to the current checkout of the project.

This should not be needed, then `find_package(Zephyr)` will always prefer to use the Zephyr within the same workspace as your application.
Only out-of-workspace applications need some extra knowledge in case you have multiple versions of Zephyr installed.

Read more here: https://docs.zephyrproject.org/latest/build/zephyr_cmake_package.html#zephyr-cmake-package-search-order

> In multi_image.cmake is a recursive invocation of CMake, which then opens the CMakeLists.txt of MCUboot, which tries to load the Zephyr module by:

Are you perhaps using the [nRF Connect SDK](https://github.com/nrfconnect/sdk-nrf) (which is including Zephyr) and not a pure Zephyr ?

Zephyr itself has no `multi_image.cmake` file.

> ```
> find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
> project(NONE)
> ```
> If that variable is requires

I assume you refer to `$ENV{ZEPHYR_BASE}` ?
And no, that variable is not required, notice the `HINTS` in front (see also https://cmake.org/cmake/help/latest/command/find_package.html#config-mode-search-procedure).

The find_package will look for Zephyr as described in the Zephyr doc link above, but **if** `ZEPHYR_BASE` is set in environment, then that setting will be honored and used.
The reason for `ZEPHYR_BASE` as environment variable is partly for historic reasons / backward compatibility from a time when ZEPHYR_BASE was required to be set, but also to allow users to explicitly point to a given Zephyr installation.
Reasons could be just for a desire to have explicit control or for test purposes. Personally I never set Zephyr base in environment, as there is no need.

**Reply by TorstenRobitzki** [reply]
Hi @tejlmand,
thank you very much for your reply! :-) Indeed, I forgot to mention that I'm using the Nordic flavor of Zephyr (when I wrote the above, I was not even aware, that this is a Nordic flavor-only feature). 

I have a single project repository, that contains a `west.yml` file that mainly contains a reference to `sdk-nrf`. The hint, that I give CMake to find the Zephyr package is relative to the CMakeLists.txt file itself, which allows me to checkout the repository at any location without setting the `ZEPHYR_BASE` environment variable:

        find_package(Zephyr REQUIRED HINTS ${CMAKE_CURRENT_LIST_DIR}/../zephyr)

The multi image build then fails, due to `ZEPHYR_BASE` not being set, which comes to a surprise, as it seems to be part of Zephyr (or the CMake Zephyr package):

    === child image b0n - CPUNET (inherited) begin ===
    loading initial cache file /Users/todi/yyyyyyyyy/zephyr-workspace/xxxxxxx/build/hci_ipc/b0n/child_image_preload.cmake
    CMake Error at CMakeLists.txt:9 (find_package):
      Could not find a package configuration file provided by "Zephyr" with any
      of the following names:

        ZephyrConfig.cmake
        zephyr-config.cmake

      Add the installation prefix of "Zephyr" to CMAKE_PREFIX_PATH or set
      "Zephyr_DIR" to a directory containing one of the above files.  If "Zephyr"
      provides a separate development package or SDK, be sure it has been
      installed.


    -- Configuring incomplete, errors occurred!
    CMake Error at /Users/todi/yyyyyyyyy/zephyr-workspace/nrf/cmake/multi_image.cmake:450 (message):
      CMake generation for b0n failed, aborting.  Command: 1
    Call Stack (most recent call first):
      /Users/todi/yyyyyyyyy/zephyr-workspace/nrf/cmake/multi_image.cmake:178 (add_child_image_from_source)
      /Users/todi/yyyyyyyyy/zephyr-workspace/nrf/samples/CMakeLists.txt:13 (add_child_image)


    -- Configuring incomplete, errors occurred!
    CMake Error at /Users/todi/yyyyyyyyy/zephyr-workspace/nrf/cmake/multi_image.cmake:450 (message):
      CMake generation for hci_ipc failed, aborting.  Command: 1
    Call Stack (most recent call first):
      /Users/todi/yyyyyyyyy/zephyr-workspace/nrf/cmake/multi_image.cmake:178 (add_child_image_from_source)
      /Users/todi/yyyyyyyyy/zephyr-workspace/nrf/samples/CMakeLists.txt:148 (add_child_image)

My workaround for this, is to set the environment variable to the required hint:

    # add_child_image_from_source() in multi_image.cmake creates sub processes running
    # cmake to build other images, which might fail to find the Zephyr package if this
    # environment variable is not set.
    set(ENV{ZEPHYR_BASE} ${CMAKE_CURRENT_LIST_DIR}/../zephyr)

    find_package(Zephyr REQUIRED HINTS ${CMAKE_CURRENT_LIST_DIR}/../zephyr)

Which works fine for me.
